### PR TITLE
ButtonLink: new component forked from Button

### PR DESCRIPTION
**Pull request by eniomoura**
#### What changed?

Added the ButtonLink component with role='link' functionality equivalent from Button component, decoupling that variant from the original component.

#### Why?

Allowing development for Button and ButtonLink separately, as a proper Button versus ButtonLink which is actually an anchor styled as a Button.

### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-6165)

### Checklist

- [ Yes ] Updated unit and Flow Tests
- [ Yes ] Updated documentation + accessibility tests
- [ Yes ] Verified accessibility: keyboard & screen reader interaction
- [ Yes ] Checked dark mode, responsiveness, and right-to-left support
- [ Yes ] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt/src/ButtonLink.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  accessibilityLabel?: string,
+  color?:
+    | 'gray'
+    | 'red'
+    | 'blue'
+    | 'transparent'
+    | 'semiTransparentWhite'
+    | 'transparentWhiteText'
+    | 'white',
+  dataTestId?: string,
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:48]: onClick was removed, bring it back with all the navigation logic

File: packages/gestalt/src/Button.js [hunk 2]
```diff
@@ -236,47 +221,6 @@ const ButtonWithForwardRef: AbstractComponent<unionProps, unionRefs> = forwardRe
         })
       : undefined;
 
-  const handleLinkClick = ({
```
Comment (AlbertCarreras) [inline packages/gestalt/src/Button.js:239]: don;t kill all this logic

**Comment by AlbertCarreras** [comment]
Button's onClick now is simpler, heads up
Screenshot by Dropbox Capture

**Comment by AlbertCarreras** [comment]
VIDEO REVIEW  https://capture.dropbox.com/xmaAzriSX2hGoIan

File: packages/gestalt/src/ButtonLink.js (hunk 1 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:50]: DONE

File: packages/gestalt/src/Button.js (hunk 2 above)
Comment (eniomoura) [reply packages/gestalt/src/Button.js]: DONE

File: packages/gestalt/src/ButtonLink.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+   * Default Buttons are sized by the text within the Button whereas full-width Buttons expand to the full width of their container.
+   */
+  tabIndex?: -1 | 0,
+  /**
+   * Callback invoked when the user clicks (press and release) on Button with the mouse or keyboard. Required with `role="button"` or `type="button"` Buttons.
+     See [GlobalEventsHandlerProvider](/web/utilities/globaleventshandlerprovider#Link-handlers) to learn more about link navigation.
+   */
+  onClick?: ({|
+    event:
+      | SyntheticMouseEvent<HTMLButtonElement>
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:73]: this is not a button anymore, just anchors

File: packages/gestalt/src/ButtonLink.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+   */
+  href: string,
+  /**
+   * Provides hints for SEO. See the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#rel) to learn more.
+   */
+  rel?: 'none' | 'nofollow',
+  /**
+   * Indicates the browsing context where an href will be opened.
+   */
+  target?: Target,
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:98]: inline target

File: packages/gestalt/src/ButtonLink.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+const SIZE_NAME_TO_PIXEL = {
+  sm: 10,
+  md: 12,
+  lg: 12,
+};
+
+type Target = null | 'self' | 'blank';
+
+type ButtonType = {|
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:31]: can we rename to Props?

File: packages/gestalt/src/ButtonLink.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    dangerouslyDisableOnNavigation: () => void,
+  ) =>
+    onClick
+      ? onClick({
+          event,
+          dangerouslyDisableOnNavigation: dangerouslyDisableOnNavigation ?? (() => {}),
+        })
+      : undefined;
+
+  const handleLinkClick = ({
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:155]: can we merge handleLinkClick & handleClick before they were used sepaartrely not anymore in this component

File: docs/pages/visual-test/ButtonLink.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// @flow strict
+import { type Node } from 'react';
```
Comment (AlbertCarreras) [inline docs/pages/visual-test/ButtonLink.js:2]: can we add a darkmode one?

File: docs/docs-components/data/components.js [hunk 8]
```diff
@@ -383,6 +383,32 @@ const componentData: $ReadOnlyArray<ComponentData> = [
       },
     },
   },
+  {
+    id: 'ButtonLink',
+    platform: {
+      web: {
+        name: 'ButtonLink',
+        visual: {
+          svg: <Button />,
```
Comment (AlbertCarreras) [inline docs/docs-components/data/components.js:392]: can we  ask in #gestalt-team to make a BbuttonLink SVG?

File: packages/gestalt/src/ButtonLink.js (hunk 3 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:73]: DONE

File: packages/gestalt/src/ButtonLink.js (hunk 4 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:98]: DONE

File: packages/gestalt/src/ButtonLink.js (hunk 5 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:31]: DONE

File: packages/gestalt/src/ButtonLink.js (hunk 6 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:155]: DONE

**Comment by AlbertCarreras** [comment]
Can we move this to ButtonLink

Screenshot 2023-07-26 at 12 27 11 PM

File: packages/gestalt/src/ButtonLink.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/ButtonLink.js:98]: Screenshot by Dropbox Capture

**Comment by AlbertCarreras** [comment]
Missing the defaults, you have to destructure the props in the arguments 

Screenshot by Dropbox Capture

File: docs/pages/visual-test/ButtonLink.js (hunk 7 above)
Comment (eniomoura) [reply docs/pages/visual-test/ButtonLink.js:2]: DONE

**Comment by AlbertCarreras** [comment]
When are you goig to write the codemod to replace Button > ButtonLink?  A follow up PR?

File: packages/gestalt/src/ButtonLink.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+   * Indicates the browsing context where an href will be opened.
+   */
+  target?: Target,
+|};
+
+const ButtonLinkWithForwardRef: AbstractComponent<ButtonType, HTMLAnchorElement> = forwardRef<
+  ButtonType,
+  HTMLAnchorElement,
+>(function ButtonLink(props: ButtonType, ref): Node {
+  const {
```
Comment (AlbertCarreras) [inline packages/gestalt/src/ButtonLink.js:105]: move this to line 104 so defaults show up in PropTable

File: docs/pages/web/button.js [hunk 10]
```diff
@@ -636,6 +636,9 @@ To control focus or position anchored components relative to Button, use \`ref\`
       <MainSection name="Related">
         <MainSection.Subsection
           description={`
+**[ButtonLink](/web/buttonlink)**
```
[the remark this answers is not in the document]
Comment (eniomoura) [reply docs/pages/web/button.js:639]: DONE

File: docs/docs-components/data/components.js (hunk 8 above)
Comment (eniomoura) [reply docs/docs-components/data/components.js:392]: Asked, waiting for answer

**Comment by eniomoura** [comment]
> When are you goig to write the codemod to replace Button > ButtonLink? A follow up PR?

Yes, on a follow-up major version that will deprecate role='link'

File: packages/gestalt/src/ButtonLink.js (hunk 9 above)
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:105]: DONE

File: packages/gestalt/src/ButtonLink.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eniomoura) [reply packages/gestalt/src/ButtonLink.js:98]: DONE

**Comment by AlbertCarreras** [comment]
There's something broken in Darkmode 

This is Button
Screenshot by Dropbox Capture

This is your ButtonLink
Screenshot 2023-07-26 at 2 14 10 PM

File: docs/pages/web/button.js [hunk 11]
```diff
@@ -34,6 +32,15 @@ const PREVIEW_HEIGHT = 300;
 export default function DocsPage({ generatedDocGen }: DocType): Node {
   return (
     <Page title={generatedDocGen?.displayName}>
+      <SlimBanner
+        message={"If you're intend to use the Button for a link, see "}
```
Comment (AlbertCarreras) [inline docs/pages/web/button.js:36]: If you intend to use the Button as a link, see

File: docs/pages/web/button.js [hunk 12]
```diff
@@ -34,6 +32,15 @@ const PREVIEW_HEIGHT = 300;
...
+        type="info"
+        helperLink={{
+          text: 'ButtonLink',
+          accessibilityLabel: 'Learn more about the ButtonLink component.',
```
Comment (AlbertCarreras) [inline docs/pages/web/button.js:40]: Learn more about ButtonLink.

**Comment by AlbertCarreras** [comment]
> > When are you goig to write the codemod to replace Button > ButtonLink? A follow up PR?
> 
> Yes, on a follow-up major version that will deprecate role='link'

I'll do it as a standalone followup. You want it to replace 200 instances. hahah And for support to the community

File: docs/pages/web/button.js (hunk 12 above)
Comment (eniomoura) [reply docs/pages/web/button.js:40]: DONE

File: docs/pages/web/button.js (hunk 11 above)
Comment (eniomoura) [reply docs/pages/web/button.js:36]: Changed the message, will explore ButtonLink implementation

**Review by AlbertCarreras — APPROVED** [review]

File: docs/docs-components/data/components.js (hunk 8 above)
Comment (eniomoura) [reply docs/docs-components/data/components.js]: DONE

**Pull request merged**